README.md
# Ecommerce Redsys
> Redsys payments with Laravel

[![Latest Version on Packagist](https://img.shields.io/packagist/v/itemvirtual/ecommerce-redsys.svg?style=flat-square)](https://packagist.org/packages/itemvirtual/ecommerce-redsys)
[![Total Downloads](https://img.shields.io/packagist/dt/itemvirtual/ecommerce-redsys.svg?style=flat-square)](https://packagist.org/packages/itemvirtual/ecommerce-redsys)


## Installation

You can install the package via composer:

```bash
composer require itemvirtual/ecommerce-redsys
```
Publish config (with `--force` option to update)
``` bash
php artisan vendor:publish --provider="Itemvirtual\EcommerceRedsys\EcommerceRedsysServiceProvider" --tag=config
```
Add this environment variable to your `.env`
``` bash
ECOMMERCE_REDSYS_ENVIRONMENT=test # or live
ECOMMERCE_REDSYS_KEY="sq7HjrUOBfKmC576ILgskD5srU870gJ7"
ECOMMERCE_REDSYS_CODE="999008881"
ECOMMERCE_REDSYS_TERMINAL=1
ECOMMERCE_REDSYS_CURRENCY=978
ECOMMERCE_REDSYS_LANGUAGE="001"
ECOMMERCE_REDSYS_NOTIFICATION_URL="${APP_URL}/redsys-notification-url"
ECOMMERCE_REDSYS_URL_OK="${APP_URL}/redsys-ok-url"
ECOMMERCE_REDSYS_URL_KO="${APP_URL}/redsys-ko-url"
ECOMMERCE_REDSYS_TRADE_NAME="${APP_NAME}"
ECOMMERCE_REDSYS_TITULAR="Your company name"
```
Add your `ECOMMERCE_REDSYS_NOTIFICATION_URL` to your `app/Http/Middleware/VerifyCsrfToken` middleware
```php
protected $except = [
    'redsys-notification-url'
];
```

## Usage
### To create the redsys payment form
```php
use Itemvirtual\EcommerceRedsys\EcommerceRedsys;

$EcommerceRedsys = new EcommerceRedsys();
$redsysForm = $EcommerceRedsys->setAmount(100)
    ->setMerchantData('YourMerchantData')
    ->setDescription('Purchase title')
    ->setSubmitButtonTitle('Pay button text')
    ->createForm();
```
Available methods
```php
$EcommerceRedsys = new EcommerceRedsys();
$params = $EcommerceRedsys->setAmount(100)
    ->setOrder(time())
    ->setMerchantData('YourMerchantData')
    ->setDescription('Purchase title')
    ->setMerchantcode('Your Custom Data')
    ->setCurrency(999)
    ->setLanguage('Your Custom Data')
    ->setTransactiontype('Your Custom Data')
    ->setTerminal(1)
    ->setMethod('Your Custom Data')
    ->setNotificationUrl('Your Custom Data')
    ->setUrlOk('Your Custom Data')
    ->setUrlKo('Your Custom Data')
    ->setTradeName('Your Custom Data')
    ->setTitular('Your Custom Data')
    ->setEnvironment('test')
    ->setSubmitButtonName('Your Custom Button title')
    ->setSubmitButtonTitle('Your Custom Button title')
    ->setSubmitButtonId('Your Custom Button title')
    ->setSubmitButtonClass('Your Custom Button title')
    ->debugParameters();
```

### To get redsys response
```php
use Itemvirtual\EcommerceRedsys\EcommerceRedsys;

$EcommerceRedsys = new EcommerceRedsys();
$parameters = $EcommerceRedsys->getMerchantParameters($request->get('Ds_MerchantParameters'));
$yourMerchantData = $EcommerceRedsys->getMerchantData($request->get('Ds_MerchantParameters'));

// validate payment
$isValid = $EcommerceRedsys->checkValidPayment($request->all());
```

### Changelog

Please see [CHANGELOG](CHANGELOG.md) for more information what has changed recently.

## Contributing

Please see [CONTRIBUTING](CONTRIBUTING.md) for details.

## Credits

-   [Itemvirtual](https://github.com/itemvirtual)
-   [All Contributors](../../contributors)

## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.

## Links

[Redsys Entorno de pruebas](https://pagosonline.redsys.es/entornosPruebas.html)


CHANGELOG.md
# Changelog

All notable changes to `ecommerce-redsys` will be documented in this file

## 1.1.0 - 2023-01-18

- Add laravel 9 support

## 1.0.3 - 2022-05-28

- Add setOrder

## 1.0.2 - 2022-04-19

- Thousands separator

## 1.0.1 - 2022-02-10

- Submit button options

## 1.0.0 - 2022-02-10

- initial release


src/EcommerceRedsys.php
<?php

namespace Itemvirtual\EcommerceRedsys;

use Sermepa\Tpv\Tpv;
use Illuminate\Support\Facades\Log;

class EcommerceRedsys
{
    private $redsys;
    private $environment;
    private $amount;
    private $order;
    private $merchantCode;
    private $currency;
    private $language;
    private $transactionType;
    private $terminal;
    private $method;
    private $notificationUrl;
    private $urlOk;
    private $urlKo;
    private $tradeName;
    private $titular;
    private $description;
    private $merchantData;

    private $submitButtonName = "submit_ecommerce_redsys";
    private $submitButtonId = "submit-ecommerce-redsys";
    private $submitButtonTitle = "Enviar";
    private $submitButtonClass = "btn btn-primary";

    const TIPO_PAGO_AUTORIZACION = 0;
    const TIPO_PAGO_PREAUTORIZACION = 1;
    const TIPO_PAGO_CONFIRMACION = 2;
    const TIPO_PAGO_DEVOLUCION_AUTOMATICA = 3;
    const TIPO_PAGO_PAGO_REFERENCIA = 4;
    const TIPO_PAGO_TRANSACCION_RECURRENTE = 5;
    const TIPO_PAGO_TRANSACCION_SUCESIVA = 6;
    const TIPO_PAGO_AUTENTICACION = 7;
    const TIPO_PAGO_CONFIRMACION_AUTENTICACION = 8;
    const TIPO_PAGO_ANULACION_PREAUTORIZACION = 9;


    public function __construct()
    {
        $this->createTransaction();
    }

    public function createTransaction()
    {
        $env = config('ecommerce-redsys.environment');

        $this->redsys = new Tpv();
        $this->redsys->setOrder(time());
        $this->redsys->setMerchantcode(config('ecommerce-redsys.parameters.' . $env . '.merchant_code'));
        $this->redsys->setCurrency(config('ecommerce-redsys.parameters.' . $env . '.currency'));
        $this->redsys->setLanguage(config('ecommerce-redsys.parameters.' . $env . '.language'));
        $this->redsys->setTransactiontype(self::TIPO_PAGO_AUTORIZACION);
        $this->redsys->setTerminal(config('ecommerce-redsys.parameters.' . $env . '.terminal'));
        $this->redsys->setMethod('C'); //Solo pago con tarjeta, no mostramos iupay
        $this->redsys->setNotification(config('ecommerce-redsys.parameters.' . $env . '.notification_url'));
        $this->redsys->setUrlOk(config('ecommerce-redsys.parameters.' . $env . '.url_ok'));
        $this->redsys->setUrlKo(config('ecommerce-redsys.parameters.' . $env . '.url_ko'));
        $this->redsys->setVersion('HMAC_SHA256_V1');
        $this->redsys->setTradeName(config('ecommerce-redsys.parameters.' . $env . '.trade_name'));
        $this->redsys->setTitular(config('ecommerce-redsys.parameters.' . $env . '.titular'));
        $this->redsys->setEnvironment(config('ecommerce-redsys.environment'));

        return $this;
    }

    private function createSignature()
    {
        $signature = $this->redsys->generateMerchantSignature(config('ecommerce-redsys.parameters.' . config('ecommerce-redsys.environment') . '.key'));
        $this->redsys->setMerchantSignature($signature);
    }

    /**
     * Get the Redsys Form
     *
     * @return mixed
     */
    public function createForm()
    {
        $this->createSignature();
        $this->redsys->setAttributesSubmit($this->submitButtonName, $this->submitButtonId, $this->submitButtonTitle, '', $this->submitButtonClass);

        return $this->redsys->createForm();
    }

    /**
     * View Redsys parameters before encrypting the form data
     *
     * @return mixed
     */
    public function debugParameters()
    {
        $parameters = $this->redsys->getParameters();
        Log::channel('ecommerce-redsys')->info('EcommerceRedsys debugParameters');
        Log::channel('ecommerce-redsys')->info(json_encode($parameters));
        return $parameters;
    }

    // *********************************************************************************** TPV RESPONSE

    /**
     * @param $Ds_MerchantParameters
     * @return null
     */
    public function getMerchantParameters($Ds_MerchantParameters)
    {
        try {
            $parameters = $this->redsys->getMerchantParameters($Ds_MerchantParameters);
            return $parameters;
        } catch (\Sermepa\Tpv\TpvException $exception) {
            Log::channel('ecommerce-redsys')->error(json_encode($exception));
        }
        return null;
    }

    /**
     * @param $Ds_MerchantParameters
     * @return mixed|null
     */
    public function getMerchantData($Ds_MerchantParameters)
    {
        try {
            $parameters = $this->redsys->getMerchantParameters($Ds_MerchantParameters);
            if (array_key_exists('Ds_MerchantData', $parameters)) {
                return $parameters['Ds_MerchantData'];
            }
        } catch (\Sermepa\Tpv\TpvException $exception) {
            Log::channel('ecommerce-redsys')->error(json_encode($exception));
        }
        return null;
    }

    /**
     * @param $Ds_MerchantParameters
     * @return int|mixed
     */
    public function getDsResponse($Ds_MerchantParameters)
    {
        try {
            $parameters = $this->redsys->getMerchantParameters($Ds_MerchantParameters);
            if (array_key_exists('Ds_Response', $parameters)) {
                return $parameters['Ds_Response'];
            }
        } catch (\Sermepa\Tpv\TpvException $exception) {
            Log::channel('ecommerce-redsys')->error(json_encode($exception));
        }
        return 99999;
    }

    /**
     * Check if is a valid payment response
     *
     * @param $post
     * @return bool
     */
    public function checkValidPayment($post)
    {
        $key = config('ecommerce-redsys.parameters.' . config('ecommerce-redsys.environment') . '.key');
        $DsResponse = $this->getDsResponse($post['Ds_MerchantParameters']);
        $DsResponse += 0;

        if ($this->redsys->check($key, $post) && $DsResponse <= 99) {
            return true;
        }

        return false;
    }

    // *********************************************************************************** SETTERS

    /**
     * @param mixed $order
     * @return EcommerceRedsys
     */
    public function setOrder($order)
    {
        $this->redsys->setOrder($order);
        return $this;
    }

    /**
     * @param mixed $environment
     * @return EcommerceRedsys
     */
    public function setEnvironment($environment)
    {
        $this->redsys->setEnvironment($environment);
        return $this;
    }

    /**
     * @param mixed $amount
     * @return EcommerceRedsys
     */
    public function setAmount($amount)
    {
        $amount = number_format($amount, 2, '.', '');
        $this->redsys->setAmount($amount);
        return $this;
    }

    /**
     * @param mixed $merchantCode
     * @return EcommerceRedsys
     */
    public function setMerchantCode($merchantCode)
    {
        $this->redsys->setMerchantcode($merchantCode);
        return $this;
    }

    /**
     * @param mixed $currency
     * @return EcommerceRedsys
     */
    public function setCurrency($currency)
    {
        $this->redsys->setCurrency($currency);
        return $this;
    }

    /**
     * @param mixed $language
     * @return EcommerceRedsys
     */
    public function setLanguage($language)
    {
        $this->redsys->setLanguage($language);
        return $this;
    }

    /**
     * @param mixed $transactionType
     * @return EcommerceRedsys
     */
    public function setTransactionType($transactionType)
    {
        $this->redsys->setTransactiontype($transactionType);
        return $this;
    }

    /**
     * @param mixed $terminal
     * @return EcommerceRedsys
     */
    public function setTerminal($terminal)
    {
        $this->redsys->setTerminal($terminal);
        return $this;
    }

    /**
     * @param mixed $method
     * @return EcommerceRedsys
     */
    public function setMethod($method)
    {
        $this->redsys->setMethod($method);
        return $this;
    }

    /**
     * @param mixed $notificationUrl
     * @return EcommerceRedsys
     */
    public function setNotificationUrl($notificationUrl)
    {
        $this->redsys->setNotification($notificationUrl);
        return $this;
    }

    /**
     * @param mixed $urlOk
     * @return EcommerceRedsys
     */
    public function setUrlOk($urlOk)
    {
        $this->redsys->setUrlOk($urlOk);
        return $this;
    }

    /**
     * @param mixed $urlKo
     * @return EcommerceRedsys
     */
    public function setUrlKo($urlKo)
    {
        $this->redsys->setUrlKo($urlKo);
        return $this;
    }

    /**
     * @param mixed $tradeName
     * @return EcommerceRedsys
     */
    public function setTradeName($tradeName)
    {
        $this->redsys->setTradeName($tradeName);
        return $this;
    }

    /**
     * @param mixed $titular
     * @return EcommerceRedsys
     */
    public function setTitular($titular)
    {
        $this->redsys->setTitular($titular);
        return $this;
    }

    /**
     * @param mixed $description
     * @return EcommerceRedsys
     */
    public function setDescription($description)
    {
        $this->redsys->setProductDescription($description);
        return $this;
    }

    /**
     * @param mixed $merchantData
     * @return EcommerceRedsys
     */
    public function setMerchantData($merchantData)
    {
        $this->redsys->setMerchantData($merchantData);
        return $this;
    }

    /**
     * @param string $submitButtonTitle
     * @return EcommerceRedsys
     */
    public function setSubmitButtonTitle(string $submitButtonTitle)
    {
        $this->submitButtonTitle = $submitButtonTitle;
        return $this;
    }

    /**
     * @param string $submitButtonName
     * @return EcommerceRedsys
     */
    public function setSubmitButtonName(string $submitButtonName)
    {
        $this->submitButtonName = $submitButtonName;
        return $this;
    }

    /**
     * @param string $submitButtonId
     * @return EcommerceRedsys
     */
    public function setSubmitButtonId(string $submitButtonId)
    {
        $this->submitButtonId = $submitButtonId;
        return $this;
    }

    /**
     * @param string $submitButtonClass
     * @return EcommerceRedsys
     */
    public function setSubmitButtonClass(string $submitButtonClass)
    {
        $this->submitButtonClass = $submitButtonClass;
        return $this;
    }

}


src/EcommerceRedsysServiceProvider.php
<?php

namespace Itemvirtual\EcommerceRedsys;

use Illuminate\Support\ServiceProvider;

class EcommerceRedsysServiceProvider extends ServiceProvider
{
    /**
     * Bootstrap the application services.
     */
    public function boot()
    {
        if ($this->app->runningInConsole()) {
            $this->publishes([
                __DIR__ . '/../config/config.php' => config_path('ecommerce-redsys.php'),
            ], 'config');

            // Registering package commands.
            // $this->commands([]);
        }
    }

    /**
     * Register the application services.
     */
    public function register()
    {
        // Automatically apply the package configuration
        $this->mergeConfigFrom(__DIR__ . '/../config/config.php', 'ecommerce-redsys');
        // Add ecommerce-redsys log channel to project
        $this->mergeConfigFrom(__DIR__ . '/../config/log-channel.php', 'logging.channels');
    }
}


config/config.php
<?php


return [

    /*
    |--------------------------------------------------------------------------
    | Redsys
    |--------------------------------------------------------------------------
    |
    | Nothing yet
    |
    */

    'environment' => env('ECOMMERCE_REDSYS_ENVIRONMENT', 'test'),

    'parameters' => [
        'test' => [
            'key' => env('ECOMMERCE_REDSYS_KEY', 'sq7HjrUOBfKmC576ILgskD5srU870gJ7'),
            'merchant_code' => env('ECOMMERCE_REDSYS_CODE', '999008881'),
            'terminal' => env('ECOMMERCE_REDSYS_TERMINAL', 1),
            'currency' => env('ECOMMERCE_REDSYS_CURRENCY', 978),
            'notification_url' => env('ECOMMERCE_REDSYS_NOTIFICATION_URL', null),
            'url_ok' => env('ECOMMERCE_REDSYS_URL_OK', false),
            'url_ko' => env('ECOMMERCE_REDSYS_URL_KO', false),
            'trade_name' => env('ECOMMERCE_REDSYS_TRADE_NAME', 'Ecommerce Redsys Shop'),
            'titular' => env('ECOMMERCE_REDSYS_TITULAR', 'Your Name'),
            'language' => env('ECOMMERCE_REDSYS_LANGUAGE', '001'),
        ],

        'live' => [
            'key' => env('ECOMMERCE_REDSYS_KEY', null),
            'merchant_code' => env('ECOMMERCE_REDSYS_CODE', null),
            'terminal' => env('ECOMMERCE_REDSYS_TERMINAL', 1),
            'currency' => env('ECOMMERCE_REDSYS_CURRENCY', 978),
            'notification_url' => env('ECOMMERCE_REDSYS_NOTIFICATION_URL', null),
            'url_ok' => env('ECOMMERCE_REDSYS_URL_OK', null),
            'url_ko' => env('ECOMMERCE_REDSYS_URL_KO', null),
            'trade_name' => env('ECOMMERCE_REDSYS_TRADE_NAME', null),
            'titular' => env('ECOMMERCE_REDSYS_TITULAR', null),
            'language' => env('ECOMMERCE_REDSYS_LANGUAGE', '001'),
        ]
    ],

];